source/async_stack.cpp

```diff
@@ -39,10 +42,47 @@ extern "C" {
 inline pthread_key_t folly_async_stack_root_tls_key = 0xFFFF'FFFFu;
 }
 
-#endif  // FOLLY_ASYNC_STACK_ROOT_USE_PTHREAD
+#else
+
+#include <vector>
+
+// TODO: expose array of AsyncStackRootHolder
+
+#endif  // ASYNC_STACK_ROOT_USE_PTHREAD
 
 namespace unifex {
 
+#if ASYNC_STACK_ROOT_USE_PTHREAD == 0
+static std::unique_ptr<std::mutex> asyncStackRootsMutex = std::make_unique<std::mutex>();
+static std::vector<void*> asyncStackRoots;
+
+static void addAsyncStackRootHolder(void* holder) noexcept {
+  std::lock_guard<std::mutex> lock(*asyncStackRootsMutex);
+  asyncStackRoots.push_back(holder);
+}
+
+static void removeAsyncStackRootHolder(void* holder) noexcept {
+  if (asyncStackRootsMutex == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(*asyncStackRootsMutex);
+  auto it = std::find(asyncStackRoots.begin(), asyncStackRoots.end(), holder);
+  if (it != asyncStackRoots.end()) {
+    asyncStackRoots.erase(it);
```

Comment:
You can do this more efficiently than `.erase()` because you don't care about order.
```suggestion
    std::swap(*it, asyncStackRoots.back());
    asyncStackRoots.pop_back();
```